src/main/java/net/neoforged/neoforge/transfer/resources/ResourceStack.java

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.transfer.resources;
+
+import com.mojang.serialization.Codec;
+import com.mojang.serialization.codecs.RecordCodecBuilder;
+import java.util.function.UnaryOperator;
+import net.minecraft.network.FriendlyByteBuf;
+import net.minecraft.network.codec.ByteBufCodecs;
+import net.minecraft.network.codec.StreamCodec;
+import net.minecraft.util.ExtraCodecs;
+import net.minecraft.world.item.ItemStack;
+import net.neoforged.neoforge.common.util.NeoForgeExtraCodecs;
+import net.neoforged.neoforge.fluids.FluidStack;
+import net.neoforged.neoforge.transfer.IStackFactory;
+import net.neoforged.neoforge.transfer.ResourceHandlerUtil;
+import net.neoforged.neoforge.transfer.TransferPreconditions;
+
+/**
+ * Represents an immutable {@link IResource} and an amount.
+ * Can be seen as an immutable version of {@link ItemStack} or {@link FluidStack}.
+ */
+public final class ResourceStack<T extends IResource<T>> {
```

Comment:
Given all the plumbing required for this class to be a drop-in replacement for ItemStack with the same level of convenience, which it can never achieve, many maintainers (at least @covers1624 @pupnewfster @Shadows-of-Fire @shartte and myself) agreed that we should keep `ResourceStack` **as a trivial record** in the first version of the transfer rework. This means that it should be trimmed to a trivial record in this slice, and **will not** be expanded in a future slice. After the whole rework is complete and released, we can look into expanding it again based on real-world feedback.